refactor(swarm)!: remove `handler` from `NetworkBehaviourAction::Dial`

thomaseizinger opened the pull request:
## Description

We create the `ConnectionId` for the new connection as part of `DialOpts`. This allows `NetworkBehaviour`s to accurately track state regarding their own dial attempts.

This patch is the main enabler of https://github.com/libp2p/rust-libp2p/pull/3254. Removing the `handler` field will allow us to deprecate the `NetworkBehaviour::new_handler` function in favor of four new ones that give more control over the connection lifecycle.

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

## Notes

- Extracted out of https://github.com/libp2p/rust-libp2p/pull/3254.
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3327.
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3335
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3372

Please review this design critically. I believe it has value on its own but at the end of the day, we only do this for https://github.com/libp2p/rust-libp2p/pull/3254.

<!-- Any notes or remarks you'd like to make about the PR. -->

## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Use single quotes to avoid running commands

thomaseizinger committed:
Use intermediary variable

thomaseizinger committed:
Make connection IDs globally unique

thomaseizinger committed:
Deprecate old new function

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Merge branch 'master' into no-run-title-as-command

thomaseizinger committed:
Remove `handler` field from `NetworkBehaviourAction::Dial`

thomaseizinger committed:
Port dcutr protocol

thomaseizinger committed:
Port relay protocol

thomaseizinger committed:
Add changelog entry to swarm

thomaseizinger committed:
Merge branch 'no-run-title-as-command' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -248,177 +233,13 @@ pub enum NetworkBehaviourAction<
+    /// of its own dial request.
+    Dial {
+        opts: DialOpts,
+        connection_id: ConnectionId,
```
Why not make `connection_id` part of `DialOpts`? For those that don't need the `connection_id`, they don't need to bother creating one, as the `DialOpts` builder would create one for them.

Do you favor being explicit, i.e. requiring the user to create the `ConnectionId`, thus being aware of it?

thomaseizinger replied:
We could move it into `DialOpts` I guess, that would make things slightly more ergonomic. Together with https://github.com/libp2p/rust-libp2p/pull/3221, we could then make the constructor `pub(crate)` and thus prevent that users assign existing connection IDs to new connections.

thomaseizinger replied:
Addressed. I also created https://github.com/libp2p/rust-libp2p/pull/3335 which I temporarily merged into this PR to keep going here.

thomaseizinger committed:
Remove ordering functionality from `ConnectionId`

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Remove unnecessary visibility qualifiers

thomaseizinger committed:
Convert `DialOpts` to struct variant

thomaseizinger committed:
Eagerly set defaults in `DialOpts`

Instead of carrying around an enum and parsing its contents later,
directly set the correct values as we construct the final `DialOpts`.

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Make `ConnectionId` part of `DialOpts`

thomaseizinger committed:
Inline `_dial`

thomaseizinger committed:
Rustfmt

thomaseizinger committed:
Fix wrong default

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger commented on protocols/dcutr/src/handler.rs:
```diff
@@ -20,61 +20,27 @@
+        if endpoint.is_relayed() {
+            Either::Left(relayed::Handler::new(endpoint.clone()))
+        } else {
+            Either::Right(direct::Handler::default()) // This is a direct connection. What we don't know is whether it is the one we created or another one that happened accidentally.
```
This is a bit of a hack that will go away once https://github.com/libp2p/rust-libp2p/pull/3254 merges. That will allow us to create the correct handler directly in the behaviour where we have the knowledge, which dials we initiated ourselves.

thomaseizinger commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -215,24 +239,34 @@ impl NetworkBehaviour for Behaviour {
+            Either::Left(_) => connection_id,
+            Either::Right(_) => match self.direct_to_relayed_connections.get(&connection_id) {
+                None => {
+                    // If the connection ID is unknown to us, it means we didn't create it so ignore any event coming from it.
```
Same here. This will be handled by the type-system once we have https://github.com/libp2p/rust-libp2p/pull/3254. Then we can return a `dummy::ConnectionHandler` for connections that we didn't initiate ourselves and thus every event we get here is always from a connection that we initiated.

thomaseizinger committed:
Expand changelog entry

thomaseizinger committed:
Move most important changelog entry to the top

thomaseizinger committed:
Add more text to changelog entry

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -229,16 +222,8 @@ pub trait PollParameters {
-    THandler: IntoConnectionHandler,
-    TInEvent = THandlerInEvent<THandler>,
-> {
+pub enum NetworkBehaviourAction<TOutEvent, TInEvent> {
```
No more trait bounds :tada:

mergify[bot] committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Remove unnecessary `..`

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

mxinden commented on swarm/src/connection/pool.rs:
```diff
@@ -528,7 +517,11 @@ where
+    pub fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        mut new_handler_fn: impl FnMut() -> THandler,
```
Providing a function to create a handler on an as-needed basis is convenient, but using this pattern across the code-base can, in my eyes, make control flow difficult to understand. An alternative could be for the `Pool` to emit an event, and for the `Swarm` to respond by returning the event and the `Handler`. This approach is more verbose, but it does not introduce a new control flow and instead uses the existing one of passing events up and down. Do you have any other alternative in mind or do you have a strong opinion about the pattern being introduced @thomaseizinger?

thomaseizinger replied:
Let me know what you think of [f5f37ab](https://github.com/libp2p/rust-libp2p/pull/3328/commits/f5f37ab3a2f78a25bf9ecc41cea92d2e41f629bb).

thomaseizinger replied:
I had to adapt the `test_connect_disconnect_ban` test, please have a thorough look!

mxinden replied:
Thanks. This looks good to me.

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -743,10 +471,9 @@ pub struct DialFailure<'a, Handler> {
+pub struct ListenFailure<'a> {
     pub local_addr: &'a Multiaddr,
     pub send_back_addr: &'a Multiaddr,
-    pub handler: Handler,
```
Should this not include the `ConnectionId` now?

thomaseizinger replied:
It does now!

mxinden commented on protocols/relay/src/priv_client.rs:
```diff
@@ -99,7 +100,9 @@ pub struct Behaviour {
-    queued_actions: VecDeque<NetworkBehaviourAction<Event, handler::Prototype>>,
+    queued_actions: VecDeque<NetworkBehaviourAction<Event, Either<handler::In, Void>>>,
+
+    pending_handler_commands: HashMap<ConnectionId, handler::In>,
```
Entries are currently not cleaned up on `FromSwarm::DialFailure`, right?

thomaseizinger replied:
Done!

mxinden commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -68,14 +69,18 @@ pub enum Error {
+
+    direct_to_relayed_connections: HashMap<ConnectionId, ConnectionId>,
+
+    outgoing_direct_connection_attempts: HashMap<(ConnectionId, PeerId), u8>,
```
```suggestion
    /// Indexed by the [`ConnectionId`] of the relayed connection.
    outgoing_direct_connection_attempts: HashMap<(ConnectionId, PeerId), u8>,
```

thomaseizinger replied:
Thanks, this also made me discover a bug!

mxinden reviewed (commented):
I think this is a much better approach than introduced with https://github.com/libp2p/rust-libp2p/pull/2191. I consider the mechanism to be significantly easier, both for the user and our internal implementation within `libp2p-swarm`.

I have a couple of questions below.

thomaseizinger committed:
Don't use else block if we return in the previous one

thomaseizinger committed:
Remove closure in favor of spawning connection from the outside

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Add `ConnectionId` to `ListenFailure`

thomaseizinger committed:
Clean up pending commands on failure

thomaseizinger committed:
Add docs and fix bug

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Consistently use `THandlerInEvent`

thomaseizinger committed:
Use type alias

thomaseizinger committed:
Remove unnecessary indentation and "quote"-ing

thomaseizinger committed:
Fix bad changelog entry

thomaseizinger committed:
Spawn connection at right point in time

Otherwise the num-established counter is wrong.

thomaseizinger commented on swarm/src/lib.rs:
````diff
@@ -502,19 +503,13 @@ where
     /// swarm.dial("/ip6/::1/tcp/12345".parse::<Multiaddr>().unwrap());
     /// ```
     pub fn dial(&mut self, opts: impl Into<DialOpts>) -> Result<(), DialError> {
````
I opted not to return the `ConnectionId` here even though it might at first be counter-intuitive. The user already has access to it via `DialOpts` so they can get it if needed but it is a little bit more verbose.

jxs replied:
why not return it and let the user decide what to do with it? Since we have it already.

thomaseizinger replied:
If we return the ID, it gives the impression that we are creating it, which isn't true.

mxinden replied:
No strong opinion. Fine with either. Let's keep as is.

thomaseizinger committed:
Fix rustdoc

thomaseizinger committed:
One more missing rustdoc link

thomaseizinger committed:
Remove `banned_peer_connections`

We no longer add connections for banned peers to the pool so we don't
have to filter them later.

thomaseizinger committed:
Fix test

Now that we no longer add connections to banned peers, we need to:

- Reduce the number of expected connections by 1 (i.e. delete line 1954)
- Act in `Stage::BannedDial` earlier (i.e. remove the `if`)

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

jxs commented on swarm/src/connection/pool.rs:
```diff
@@ -518,13 +504,57 @@ where
+        Ok(())
+    }
+
+    pub fn spawn_connection(
```
yeah I think this fits here better, and being handled by the `Swarm` on `Swarm::handle_pool_event` after receiving `ConnectionEstablished` rather than on the `Pool` before returning it :rocket:

thomaseizinger replied:
Yeah it turned out surprisingly clean and fixes up some weird state tracking where we had to "filter" connection IDs that are banned. Now we just don't spawn the connection in the first place :)

jxs commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -148,40 +157,57 @@ impl Behaviour {
-            relayed_connection_id,
-            role: handler::Role::Initiator { attempt },
-        } = handler
+        let peer_id = if let Some(peer_id) = peer_id {
+            peer_id
+        } else {
+            return;
+        };
```
nit: I think we can do:
```suggestion
        let Some(peer_id) = peer_id {
            peer_id
        } else {
            return;
        };
```
if we bump `rust-version` on `Cargo.toml` to [`1.65`](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements) :stuck_out_tongue:

thomaseizinger replied:
I didn't want to bump MSRV just for that but I am certainly looking forward to `let-else` :)

jxs reviewed (approved):
Overall looks good to me Thomas! I am not very familiar with `dctur` and `relay`yet so Max and Elena should also chime in.

mxinden commented on core/src/muxing/boxed.rs:
```diff
@@ -13,6 +13,12 @@ pub struct StreamMuxerBox {
 
+impl fmt::Debug for StreamMuxerBox {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StreamMuxerBox").finish_non_exhaustive()
```
Fancy. Wasn't aware of `finish_non_exhaustive`.

mxinden reviewed (approved):
Ready to go from my end.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -762,74 +751,77 @@ where
-                    // Mark the connection for the banned peer as banned, thus withholding any
-                    // future events from the connection to the behaviour.
-                    self.banned_peer_connections.insert(id);
                     self.pool.disconnect(peer_id);
```
Given that we drop `connection` here, is the call to `pool.disconnect` still needed? There should be no connection to the peer in the pool. The previous `Swarm::ban_peer_id` should have already called `pool.disconnect`.

Also, if I am not mistaken, we would previously gracefully close connections to banned peers via the `pool.disconnect` call. Given that `connection` is now just dropped, this is no longer the case, right?

thomaseizinger replied:
> Given that we drop `connection` here, is the call to `pool.disconnect` still needed? There should be no connection to the peer in the pool. The previous `Swarm::ban_peer_id` should have already called `pool.disconnect`.

Originally I thought that we might have other connections to this peer but I guess that isn't actually possible because we would have disconnected those at the time of banning the peer.

> Also, if I am not mistaken, we would previously gracefully close connections to banned peers via the `pool.disconnect` call. Given that `connection` is now just dropped, this is no longer the case, right?

True. I didn't bother much because this code will go away anyway once we implement it via the new `NetworkBehaviour` callbacks but yeah, it is a change in behaviour so we should probably keep this behaviour.

This made me think though, we should probably gracefully close connections in https://github.com/libp2p/rust-libp2p/pull/3254 that have been rejected by a `NetworkBehaviour`.

thomaseizinger replied:
I've pushed some more commits, let me know what you think!

mxinden replied:
Changes look good to me, thanks!

mxinden reviewed (commented):
This still needs further investigation:

thomaseizinger committed:
Inline `spawn` function

thomaseizinger committed:
Box future inside `spawn`

thomaseizinger committed:
Merge impl blocks

thomaseizinger committed:
Gracefully close connection to banned peer

thomaseizinger committed:
Ignore error when closing connection

mergify[bot] committed:
Merge branch 'master' into 2824-no-handler-in-dial
